File: scripts/algorithms/F/Find the Longest Valid Obstacle Course at Each Position/Find the Longest Valid Obstacle Course at Each Position.py
from bisect import bisect_right
class Solution:
    def longestObstacleCourseAtEachPosition(self, obstacles: List[int]) -> List[int]:
        longest, res = [], []
        for i in range(len(obstacles)):
            idx = bisect_right(longest, obstacles[i])
            if idx == len(longest):
                longest.append(obstacles[i])
            else:
                longest[idx] = obstacles[i]
            res.append(idx+1)
        return res

File: scripts/algorithms/F/Find the Longest Valid Obstacle Course at Each Position/Find the Longest Valid Obstacle Course at Each Position.js
// Runtime: 715 ms (Top 57.14%) | Memory: 85.7 MB (Top 42.86%)
var longestObstacleCourseAtEachPosition = function(obstacles) {
    var n = obstacles.length;
    var lis = [];
    var res = new Array(n).fill(0);
    for(var i = 0; i<n; i++)
    {
        if(lis.length>0 && obstacles[i] >= lis[lis.length-1])
        {
            lis.push(obstacles[i]);
            res[i] = lis.length;
        }
        else
        {
            // find the upper bound
            var l = 0;
            var r = lis.length;
            while(l<=r)
            {
                var mid = Math.floor((l+r)/2);
                if(lis[mid]<=obstacles[i])
                {
                    l = mid+1;
                }
                else
                {
                    r = mid-1;
                }
            }
            lis[l] = obstacles[i];
            res[i] = l+1;
        }
    }
    return res;
}